src/librustc_mir/transform/util.rs

```diff
@@ -21,6 +21,31 @@ pub fn update_basic_block_ids(mir: &mut Mir, replacements: &[BasicBlock]) {
     }
 }
 
+/// Run a function on every reachable basic block, then delete any unreachable blocks.
+/// The function given should not add or remove any blocks from the control flow graph.
+pub fn adjust_reachable_basic_blocks<F: FnMut(&mut Mir, BasicBlock)>(mir: &mut Mir, mut func: F) {
+    let mut seen = vec![false; mir.basic_blocks.len()];
+
+    let mut worklist = vec![START_BLOCK];
+    while let Some(bb) = worklist.pop() {
+        func(mir, bb);
+
+        for succ in mir.basic_block_data(bb).terminator.successors() {
+            if !seen[succ.index()] {
+                seen[succ.index()] = true;
+                worklist.push(*succ);
+            }
+        }
+    }
+
+    // These blocks are always required.
```

Comment:
Don't these lines have to go before the while loop? Because right now as we enter the while loop no element in `seen` is true, and so in the for loop the if condition never holds and we never actually push anything onto the worklist.